### DISCUSSION TITLE
Theme capabilities?

### DISCUSSION THREAD

**Discussion by mdekrey** [mdx-editor/editor#317] [Q&A]
Are there current plans to support theme capabilities for use with sites where other custom CSS rules are applied?

The problems I'm having:
- A dark theme gets ignored for plain paragraph tags due to the `.contentEditable` selector's setting of `color` which is not present on `p`.
- Default styling, such as `code` blocks, get conflicting styles from Lexical vs standard CSS (visible in the current Live Editor, as the lexical class overrides the `@tailwindcss/typography` style)
- Using `@tailwindcss/typography` causes cascades to internal elements, so cannot be used in many of my advanced scenarios.

These are the parts that I think would need to change:
- Allow an override of the `LexicalComposer` `theme` configuration
- Allow override (not just merging of) class names for the styles used in `RichTextEditor`
- CSS variables set by the default styles should still be applied, but should be documented and easily overridable

A couple options that I can think of (my preferred one is first, but I'd be willing to provide a PR with either if some direction of desired conventions is provided):
- Allow a CSS Module-like object as a property to replace the current CSS Modules, using the current ones (`ui` and `lexical-theme`) as default values
- Add more props options for other classNames, including something to put the editor in "headless mode" to remove default styles
Both options would need documentation of the CSS variables to override for the editors.

I'd suggest that default styles for toolbars, editors, etc. are only affected via the CSS variables; those could be further themed via plugin options.

**Comment by petyosi** [comment]
Thank you for your input, happy to have this discussion. This is obviously a necessity to make the editor more flexible. 

To ensure that we're on the same page, I want to confirm the challenges you see are with the content styling itself - not with the toolbar/editors/etc. The "chrome" of the editor is a different story, a large one must I say. 

Up to your suggested changes:
> Allow an override of the LexicalComposer theme configuration

100% - happy to accept a PR on the matter. It needs a decent documentation, though.
> Allow override (not just merging of) class names for the styles used in RichTextEditor

This one I don't get - isn't this the same as the previous suggestion?
> CSS variables set by the default styles should still be applied, but should be documented and easily overridable

This part I'm also not entirely clear on. Do you refer [to these](https://github.com/mdx-editor/editor/blob/8c04e42a4e8483ac4297c90aade2e132f8b6d8c8/src/styles/ui.module.css#L38-L40)? More documentation is always good, but in this case I would rather just link to the source code itself. What are the challenges with overriding? Perhaps a matter of documentation. 

> Allow a CSS Module-like object as a property to replace the current CSS Modules, using the current ones (ui and lexical-theme) as default values

I think that if replacing the Lexical theme configuration is possible, then overriding the lexical-theme CSS module is redundant? As far as `ui` goes, I think replacing it with something is far from practical - people would copy its current contents, creating an unmaintainable fork. I might be wrong though.

**Reply by mdekrey** [reply]
Hmm, let me try to explain the different extreme options I was suggesting in more clarity; these options could be blended.

1. More classNames options to blend with the default classes.
    The current state has some classes that can be merged (such as via the `contentEditableClassName` prop). Extending this to all elements would allow for a custom theme by simply tweaking from the standard, but would require CSS knowledge of specificity and the existing styles to override anything other than what is supported by the current CSS variables. 

2. Full override-only of classes.
    A full override of styles could work like an object where each property was a `className`, similar to a CSS module, as well as a lexical `EditorThemeClasses` object to allow full customization of lexical (with a link to [lexical's theming docs](https://lexical.dev/docs/getting-started/theming)). By overriding styles, users could more easily get a custom theme, but it would be more difficult to make slight adjustments from the default.

A possible blended option (which I think would be both easy to maintain and upgrade to) would have these code changes:
- Keep current props. This makes it easy for current developers to continue to upgrade without redoing their own customizations.
- Add a `lexicalTheme` prop that will replace `src/styles/lexicalTheme` if provided. (For devs who want to customize what the editable content itself looks like, like using Tailwind without the `prose` class.)
- Add a prop for styles that will override the base `src/styles/ui.module.css` properties; at a minimum including these four: `contentEditable`, `placeholder`, `editorRoot`, and `editorWrapper`. (These would be for devs who want a fully custom theme to the editor itself, along with swapping out the toolbar plugin and needing to be mindful of official plugin styles.)

**Reply by mdekrey** [reply]
https://github.com/mdx-editor/editor/compare/main...mdekrey:mdx-editor:feat/theme-overrides is a quick implementation of the blended option described in my last comment. (I'm not sure if I leveraged `gurx` appropriately.) I'll happily open a PR if this looks close to what you'd want for this project!

**Reply by petyosi** [reply]
I see what you mean. 

- I'm happy to accept the `lexicalTheme` addition. 
- I understand the motivation, but I can't accept the theme override. Implementation complexity aside (theme needs to replace every occurrence of the ui.module import), this would lead to devs copy-pasting the base stylesheet, making any potential upgrades a nightmare for them and for me as an author. There has to be a better, safer way to do that.

**Reply by petrpacas** [reply]
Any news on this? I'm too very much in need of removing the default Lexical styles or resetting them somehow
